### PR TITLE
[PyTorch] Add ops for MoE grouped MLP

### PR DESCRIPTION
**Pull request by timmoon10**
# Description

This PR adds ops needed for the grouped MLP block in Mixture-of-Experts models. In particular, it adds a grouped linear op (similar to the `GroupedLinear` module) and a `ScaledSwiGLU` op. It is the same as https://github.com/NVIDIA/TransformerEngine/pull/2622, but doesn't include the fused ops with experimental kernels. Closes https://github.com/NVIDIA/TransformerEngine/issues/2560.

## Type of change

- [ ] Documentation change (change only to the documentation, either a fix or a new content)
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Infra/Build change
- [ ] Code refactoring

## Changes

- Add grouped linear op
- Add scaled SwiGLU op
- Handle edge cases in `noop_cat` function

# Checklist:

- [x] I have read and followed the [contributing guidelines](https://github.com/NVIDIA/TransformerEngine/blob/main/CONTRIBUTING.rst)
- [x] The functionality is complete
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes

### CODE REVIEW

File: transformer_engine/pytorch/ops/basic/swiglu.py [hunk 1]
```diff
@@ -0,0 +1,415 @@
...
+                grad_input = grad_input.transpose(1, 2).contiguous()
+                grad_input = grad_input.view(shape)
+
+        # Compute scales grad by recomputing SwiGLU
+        grad_extra_input = None
+        if ctx.extra_input_requires_grad:
+            swiglu_out = tex.swiglu(swiglu_in, None)
+            grad_extra_input = torch.linalg.vecdot(swiglu_out, grad_output)
+
+        # Clear input tensor if possible
```
[the remark this answers is not in the document]
Comment (timmoon10) [reply transformer_engine/pytorch/ops/basic/swiglu.py:495]: [`torch.linalg.vecdot`](https://docs.pytorch.org/docs/stable/generated/torch.linalg.vecdot.html) computes the dot product over the last tensor dim and leaves all other dims unchanged. The existing code is correct and the suggested code exposes an unnecessary memory-bound kernel. See https://github.com/NVIDIA/TransformerEngine/pull/2622#discussion_r2784590137.

File: tests/pytorch/test_fusible_ops.py [hunk 2]
```diff
@@ -1924,6 +1967,231 @@ def test_dropout(
...
+        input_requires_grad: bool,
+        weight_requires_grad: bool,
+    ) -> None:
+        """Grouped GEMM"""
+
+        # Split sizes
+        split_sizes = [split_alignment * i for i in range(group_size)]
+        random.shuffle(split_sizes)
+        split_sizes = torch.tensor(split_sizes, dtype=torch.int, device=device)
+
```
[the remark this answers is not in the document]
Comment (timmoon10) [reply tests/pytorch/test_fusible_ops.py:2012]: These split points are chosen this way precisely to test the edge case you mention. You've been making a lot of mistakes and nits while reviewing this PR, please be more thoughtful and substantive in the future.

File: transformer_engine/pytorch/ops/basic/grouped_linear.py [hunk 3]
```diff
@@ -0,0 +1,700 @@
...
+            dtype = torch.get_autocast_dtype("cuda")
+        else:
+            dtype = self.weight0.dtype
+
+        # Extract split sizes from extra input
+        split_sizes = basic_op_extra_inputs[0][0]
+        split_sizes_int = [int(s) for s in split_sizes.tolist()]
+        if len(split_sizes_int) != num_groups:
+            raise ValueError(f"Expected {num_groups} splits, but got {len(split_sizes_int)}.")
+
```
[the remark this answers is not in the document]
Comment (timmoon10) [reply transformer_engine/pytorch/ops/basic/grouped_linear.py:504]: This proposed check requires a GPU-CPU synchronization, and therefore breaks compatibility with CUDA Graphs. We currently copy the split points to CPU, but that will be removed once we support the grouped tensor API from https://github.com/NVIDIA/TransformerEngine/pull/2654.

File: tests/pytorch/test_fusible_ops.py [hunk 4]
```diff
@@ -172,6 +173,29 @@ def make_reference_and_test_tensors(
     return ref, test
 
 
+def assert_close(
```
Comment (ptrendx) [inline tests/pytorch/test_fusible_ops.py]: Shouldn't this be in some more generic place?

File: tests/pytorch/test_fusible_ops.py [hunk 5]
```diff
@@ -1727,10 +1762,18 @@ def test_swiglu(
...
 
         # Check results
-        y_test = y_test.to(dtype=torch.float64, device="cpu")
-        dx_test = x_test.grad.to(dtype=torch.float64, device="cpu")
-        torch.testing.assert_close(y_test, y_ref, **tols)
-        torch.testing.assert_close(dx_test, x_ref.grad, **tols)
+        assert_close(y_test, y_ref, **tols)
+        assert_close(x_test.grad, x_ref.grad, **tols)
+
+    def test_interleaved_swiglu(self):
```
Comment (ptrendx) [inline tests/pytorch/test_fusible_ops.py:1757]: Why a separate test rather than a parameter in the other test? Does this only work for specific scenarios?

File: transformer_engine/pytorch/module/_common.py [hunk 6]
```diff
@@ -125,14 +125,19 @@ def forward(
...
+        # Out-of-place concatenation when view tensors have different storage
+        # Note: This works around an edge case with the split_quantize
+        # function, which might allocate a buffer and construct
+        # subviews. However, in order to reduce CPU overheads, these
+        # views are configured manually outside of PyTorch. PyTorch
+        # doesn't know these views share the same memory, and it
+        # blocks us from reconstructing the full tensor because it
+        # thinks we are accessing out-of-bounds memory.
+        if tensors[0].untyped_storage().nbytes() < out_shape[dim] * data_ptr_stride:
+            return torch.cat(tensors, dim=dim)
```
Comment (ptrendx) [inline transformer_engine/pytorch/module/_common.py:129]: Should this be done before this big for loop to short circuit it?

File: transformer_engine/pytorch/ops/basic/activation.py [hunk 7]
```diff
@@ -27,8 +27,6 @@
     "SReLU",
     "SReGLU",
     "SiLU",
-    "SwiGLU",
-    "ClampedSwiGLU",
 ]
```
Comment (ptrendx) [inline transformer_engine/pytorch/ops/basic/activation.py:30]: This looks like a breaking change in the API? Or we assume that people would always use those ops from the top level?

File: transformer_engine/pytorch/ops/basic/grouped_linear.py [hunk 8]
```diff
@@ -0,0 +1,700 @@
+# Copyright (c) 2022-2026, NVIDIA CORPORATION & AFFILIATES. All rights reserved.
+#
+# See LICENSE for license information.
+
+"""Fusible operation for bias."""
```
Comment (ptrendx) [inline transformer_engine/pytorch/ops/basic/grouped_linear.py]: Not really :-).

File: transformer_engine/pytorch/ops/basic/grouped_linear.py [hunk 9]
```diff
@@ -0,0 +1,700 @@
...
+            f"and {self.num_extra_outputs} extra tensor outputs. "
+            "It overrides `fuser_backward` instead of `op_backward`."
+        )
+
+    def fuser_forward(
+        self,
+        basic_op_ctxs: list[OperationContext],
+        input_: torch.Tensor,
+        *,
+        basic_op_extra_inputs: list[tuple[torch.Tensor, ...]],
```
Comment (ptrendx) [inline transformer_engine/pytorch/ops/basic/grouped_linear.py:468]: To be honest I don't know how I feel about the decision to deal with splits as the extra input. This basically makes putting a router/permutation into the same sequential as the grouped linear impossible, right?

Comment (ptrendx) [reply transformer_engine/pytorch/ops/basic/grouped_linear.py:468]: The way I thought about it would be to handle tuple of Tensors instead. If we want to support attention in sequential then we would need to go that route anyway (in order to be able to express what is currently the fp8_mha nicely).

File: transformer_engine/pytorch/ops/basic/grouped_linear.py [hunk 10]
```diff
@@ -0,0 +1,700 @@
...
+    canonicalize_dtype,
+    clear_tensor_data,
+    devices_match,
+    round_up_to_nearest_multiple,
+)
+from .._common import is_quantized_tensor, maybe_dequantize
+from ..op import BasicOperation, OperationContext
+
+
+class GroupedLinear(BasicOperation):
```
Comment (ptrendx) [inline transformer_engine/pytorch/ops/basic/grouped_linear.py:37]: Please mark it experimental for now. I think we will need to iterate on this design.

File: transformer_engine/pytorch/ops/basic/swiglu.py [hunk 11]
```diff
@@ -0,0 +1,415 @@
+# Copyright (c) 2022-2026, NVIDIA CORPORATION & AFFILIATES. All rights reserved.
+#
+# See LICENSE for license information.
+
+"""Fusible operation for multiplying with extra input tensor."""
```
Comment (ptrendx) [inline transformer_engine/pytorch/ops/basic/swiglu.py]: Not really.

File: transformer_engine/pytorch/ops/basic/swiglu.py [hunk 12]
```diff
@@ -0,0 +1,415 @@
...
+from ...tensor import Float8CurrentScalingQuantizer, Quantizer
+from ...utils import clear_tensor_data
+from ..op import BasicOperation, OperationContext
+from .._common import maybe_dequantize
+
+__all__ = ["SwiGLU", "ClampedSwiGLU", "ScaledSwiGLU"]
+
+
+class SwiGLU(BasicOperation):
+    r"""Swish gated linear unit
```
Comment (ptrendx) [inline transformer_engine/pytorch/ops/basic/swiglu.py:24]: The doc should include the information about the interleaving parameter and what it means.

File: transformer_engine/pytorch/ops/basic/swiglu.py [hunk 13]
```diff
@@ -0,0 +1,415 @@
...
+    from GPT OSS implementation where the gates/pre-activations are assumed to be interleaved in the input tensor.
+
+    Parameters
+    ----------
+    limit : float
+        The clamp limit.
+    alpha : float
+        The scaling factor for the sigmoid function used in the activation.
+    cache_quantized_input : bool, default = ``False``
+        Quantize input tensor when caching for use in the backward pass.
```
Comment (ptrendx) [inline transformer_engine/pytorch/ops/basic/swiglu.py:208]: We will need similar interleaving support here pretty soon.

File: transformer_engine/pytorch/ops/basic/swiglu.py [hunk 14]
```diff
@@ -0,0 +1,415 @@
...
+        return dx, ()
+
+
+class ScaledSwiGLU(BasicOperation):
+    r"""SwiGLU with post-scaling
+
+    If the SwiGLU output has shape ``(d_1, ..., d_n)``, it is
+    multiplied with an extra input tensor of shape
+    ``(d_1, ..., d_{n-1})``.
+
```
Comment (ptrendx) [inline transformer_engine/pytorch/ops/basic/swiglu.py:352]: Also please add the docs for the interleaving. BTW - is the "cache quantized input" parameter only for Lora or something that it is only in the regular swiglu and nowhere else?

File: transformer_engine/pytorch/ops/basic/swiglu.py [hunk 15]
```diff
@@ -0,0 +1,415 @@
...
+                shape[-1] // (2 * self.glu_interleave_size),
+                2,
+                self.glu_interleave_size,
+            )
+            swiglu_in = swiglu_in.transpose(1, 2).contiguous()
+            swiglu_in = swiglu_in.view(shape)
+
+        # Compute scaled SwiGLU
+        swiglu_out = tex.swiglu(swiglu_in, None)
+        out = swiglu_out * scales.unsqueeze(-1)
```
Comment (ptrendx) [inline transformer_engine/pytorch/ops/basic/swiglu.py:424]: Considering it is implemented with 2 kernels anyway, what is the benefit of having this operation here? I would prefer to have the ScaleWithExtraInput basic op or something like that instead.

File: tests/pytorch/test_fusible_ops.py (hunk 5 above)
Comment (timmoon10) [reply tests/pytorch/test_fusible_ops.py:1757]: None of the other activations support interleaving since it's so hacky and specific to the fused GEMM + SwiGLU kernel.

File: transformer_engine/pytorch/ops/basic/activation.py (hunk 7 above)
Comment (timmoon10) [reply transformer_engine/pytorch/ops/basic/activation.py:30]: Accessing via `te.ops` or `te.ops.basic` is definitely the best practice. I don't see any downstream code accessing these via `te.ops.basic.activation`: https://github.com/search?q=ops.basic.activation&type=code
Megatron-LM accesses `SwiGLU` via `te.ops`:
https://github.com/NVIDIA/Megatron-LM/blob/24515084a6395bc4c30e20368d47d073868e95e6/megatron/core/extensions/transformer_engine.py#L2076
I don't see `ClampedSwiGLU` being accessed anywhere.

File: transformer_engine/pytorch/ops/basic/grouped_linear.py (hunk 9 above)
Comment (timmoon10) [reply transformer_engine/pytorch/ops/basic/grouped_linear.py:468]: It's hard to make this work well with a sequential structure.
- Supporting ops with tuple inputs/outputs is possible but messy. We'll need to add more complicated logic to handle grads properly. If the number of tensors doesn't match between two ops, maybe we could use extra inputs/outputs to make up the difference.
- Passing a tuple will allow us to pass split points to FC1, but not FC2. I suppose we could make this work if we make grouped-ness a persistent property, and make sure the other ops support grouped implementations. I worry that we would be at risk of overfitting for MoEs.

File: transformer_engine/pytorch/ops/basic/swiglu.py (hunk 14 above)
Comment (timmoon10) [reply transformer_engine/pytorch/ops/basic/swiglu.py:352]: It was an experimental feature for LoRA that we never really ended up using.

File: transformer_engine/pytorch/ops/basic/swiglu.py (hunk 15 above)
Comment (timmoon10) [reply transformer_engine/pytorch/ops/basic/swiglu.py:424]: This was the approach I used in my initial implementation (https://github.com/NVIDIA/TransformerEngine/pull/2605), but it's not compatible with the fused GEMM + SwiGLU kernel (https://github.com/NVIDIA/cudnn-frontend/blob/main/python/cudnn/grouped_gemm/grouped_gemm_swiglu/api.py). If we have a standalone scale op, then we need to cache its input for the backward pass. However, the fused kernel assumes you are doing activation recompute and it only outputs the SwiGLU input and scale output. Rather than intertwining the implementations of the SwiGLU and scale to support activation recompute, I just implemented a new op that does it explicitly

File: transformer_engine/pytorch/ops/basic/swiglu.py (hunk 13 above)
Comment (timmoon10) [reply transformer_engine/pytorch/ops/basic/swiglu.py:208]: Hopefully the fast kernels team has mercy on us.

File: tests/pytorch/test_fusible_ops.py (hunk 4 above)
Comment (timmoon10) [reply tests/pytorch/test_fusible_ops.py]: Moved to `utils.py`.

File: tests/pytorch/test_fusible_ops.py (hunk 5 above)
Comment (timmoon10) [reply tests/pytorch/test_fusible_ops.py:1757]: Also, separating the test like this is good practice. Just adding more parameters to the SwiGLU test results in exponential growth in the number of cases, while creating standalone tests results in linear growth.
